File: frontend/apps/child/src/api/learning.ts
// Shared learning API types and functions for the child app.
// NOTE: Type definitions (TopicResponse, ProgressResponse, etc.) are duplicated
// in apps/main/src/api/learning.ts — extract to @numina/types when drift becomes painful.
import http from './index'

// --- Interfaces (IDs as string — Snowflake serialization) ---

export interface TopicResponse {
  id: string
  topic_key: string
  topic_type: string
  subject: string
  domain: string | null
  name: string | null
  name_zh: string | null
  description: string
  description_zh: string | null
  age_range_start: number | null
  age_range_end: number | null
  age_group: string
  centrality: number | null
  evidence: string[]
  evidence_zh: string[] | null
  assessment_prompt: string | null
  assessment_prompt_zh: string | null
  standards: string[]
  ability_dimensions: string[] | null
  deprecated: boolean
}

export interface ProgressResponse {
  id: string
  child_id: string
  topic_id: string
  mastery_level: string
  mastery_score: number | null
  completed_via: string | null
  attempts: number
  xp_earned: number
  last_practice_at: string | null
  first_mastered_at: string | null
  stability: number | null
  next_review_at: string | null
  ability_dimensions_score: Record<string, number> | null
}

export interface AssignmentResponse {
  id: string
  family_id: string
  child_id: string
  topic_id: string
  path_id: string | null
  created_by: string
  assignment_type: string
  status: string
  priority: number
  due_date: string | null
  created_at: string
  completed_at: string | null
  topic?: TopicResponse | null
}

export interface SessionResponse {
  id: string
  assignment_id: string | null
  child_id: string
  topic_id: string
  thread_id: string | null
  session_type: string
  score: number | null
  duration_seconds: number | null
  started_at: string
  ended_at: string | null
}

export interface SessionCreate {
  topic_id: string
  assignment_id?: string
  session_type?: string
}

export interface ChildLearningOverview {
  child_id: string
  child_name: string
  mastered_count: number
  learning_count: number
  available_count: number
  locked_count: number
  review_count: number
  total_study_minutes: number
}

export interface ChildProgressOverview {
  mastered_count: number
  learning_count: number
  available_count: number
  locked_count: number
  review_count: number
  assessing_count: number
  parent_review_count: number
  total_study_minutes: number
  today_study_minutes: number
}

export interface TodayLearningResponse {
  current_topic: TopicResponse | null
  pending_assignment: AssignmentResponse | null
  recommended_topic: TopicResponse | null
  study_minutes_today: number
}

export interface TopicGraphResponse {
  topic: TopicResponse
  prerequisites: TopicResponse[]
  dependents: TopicResponse[]
}

// Composite type: progress with topic detail for the map view
export interface ProgressWithTopic extends ProgressResponse {
  topic: TopicResponse
}

// --- API Functions ---

export async function getMyLearningMap(): Promise<ProgressResponse[]> {
  const res = await http.get('/child/learning/map')
  return res.data
}

export async function getMyAssignments(status?: string): Promise<AssignmentResponse[]> {
  const res = await http.get('/child/learning/assignments', {
    params: status ? { status } : {},
  })
  return res.data
}

export async function getTopicDetail(topicId: string): Promise<TopicResponse> {
  const res = await http.get(`/child/learning/topics/${topicId}`)
  return res.data
}

export async function createSession(req: SessionCreate): Promise<SessionResponse> {
  const res = await http.post('/child/learning/sessions', req)
  return res.data
}

export async function getSession(sessionId: string): Promise<SessionResponse> {
  const res = await http.get(`/child/learning/sessions/${sessionId}`)
  return res.data
}

export async function startAssessment(sessionId: string): Promise<ProgressResponse> {
  const res = await http.post(`/child/learning/sessions/${sessionId}/start-assessment`)
  return res.data
}

export async function submitAssignment(assignmentId: string): Promise<ProgressResponse> {
  const res = await http.post(`/child/learning/assignments/${assignmentId}/submit`)
  return res.data
}

export async function getMyProgress(): Promise<ChildProgressOverview> {
  const res = await http.get('/child/learning/progress')
  return res.data
}

export async function getTopicsBatch(ids: string[]): Promise<TopicResponse[]> {
  const res = await http.get(`/learning/topics/batch?ids=${ids.join(',')}`)
  return res.data
}

export async function endSession(sessionId: string): Promise<SessionResponse> {
  const res = await http.post(`/child/learning/sessions/${sessionId}/end`)
  return res.data
}

export async function getTodayLearning(): Promise<TodayLearningResponse> {
  const res = await http.get('/child/learning/today')
  return res.data
}

export async function getTopicGraph(topicId: string): Promise<TopicGraphResponse> {
  const res = await http.get(`/learning/topics/${topicId}/graph`)
  return res.data
}

/** Translate a topic's content to Chinese. Returns translated fields. */
export async function translateTopic(topicId: string): Promise<{
  name_zh: string | null
  description_zh: string | null
  evidence_zh: string[] | null
  assessment_prompt_zh: string | null
}> {
  const res = await http.post(`/learning/topics/${topicId}/translate`)
  return res.data
}

File: server/apps/backend/app/schemas/learning.py
# server/apps/backend/app/schemas/learning.py
"""Request/response schemas for Learning OS APIs."""

from __future__ import annotations

from datetime import date, datetime

from pydantic import BaseModel, field_validator

from apps.backend.app.schemas.base import SnowflakeBase

# --- Knowledge Graph (Global) ---

class TopicResponse(SnowflakeBase):
    id: int
    topic_key: str
    topic_type: str
    subject: str
    domain: str | None
    name: str | None
    name_zh: str | None
    description: str
    description_zh: str | None
    age_range_start: int | None
    age_range_end: int | None
    age_group: str
    centrality: float | None
    evidence: list[str] = []
    evidence_zh: list[str] | None
    assessment_prompt: str | None
    assessment_prompt_zh: str | None
    standards: list[str] = []
    ability_dimensions: list[str] | None
    deprecated: bool


class TopicGraphResponse(BaseModel):
    """Local subgraph: prerequisites + dependents of a topic."""
    topic: TopicResponse
    prerequisites: list[TopicResponse]
    dependents: list[TopicResponse]


class ClusterResponse(SnowflakeBase):
    id: int
    subject: str
    domain: str
    age_range_start: int | None
    age_group: str
    summary: str


class SubjectSummary(BaseModel):
    subject: str
    topic_count: int
    mastered_count: int = 0  # filled per-child


# --- Progress (Per-Family) ---

class ProgressResponse(SnowflakeBase):
    id: int
    child_id: int
    topic_id: int
    mastery_level: str
    mastery_score: float | None
    completed_via: str | None
    attempts: int
    xp_earned: int
    last_practice_at: datetime | None
    first_mastered_at: datetime | None
    stability: float | None
    next_review_at: datetime | None
    ability_dimensions_score: dict | None


# --- Assignment ---

class AssignmentCreate(BaseModel):
    child_id: int
    topic_id: int
    assignment_type: str = "parent_assigned"
    priority: int = 0
    due_date: date | None

    @field_validator("assignment_type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        allowed = {"parent_assigned", "ai_recommended", "self_selected"}
        if v not in allowed:
            raise ValueError(f"assignment_type must be one of {allowed}")
        return v


class AssignmentResponse(SnowflakeBase):
    id: int
    family_id: int
    child_id: int
    topic_id: int
    path_id: int | None
    created_by: int
    assignment_type: str
    status: str
    priority: int
    due_date: date | None
    created_at: datetime
    completed_at: datetime | None
    topic: TopicResponse | None = None  # expanded when needed


# --- Session ---

class SessionCreate(BaseModel):
    topic_id: int
    assignment_id: int | None = None
    session_type: str = "tutorial"


class SessionResponse(SnowflakeBase):
    id: int
    assignment_id: int | None
    child_id: int
    topic_id: int
    thread_id: str | None
    session_type: str
    score: float | None
    duration_seconds: int | None
    started_at: datetime
    ended_at: datetime | None


# --- Assessment Attempt ---

class AssessmentAttemptResponse(SnowflakeBase):
    id: int
    child_id: int
    topic_id: int
    session_id: int | None
    assessment_type: str
    score: float | None
    passed: bool
    ai_confidence: float | None
    duration_seconds: int | None
    created_at: datetime


# --- Composite / Dashboard ---

class ChildLearningOverview(SnowflakeBase):
    child_id: int
    child_name: str
    mastered_count: int
    learning_count: int
    available_count: int
    locked_count: int
    review_count: int
    total_study_minutes: int


class ReviewItemResponse(SnowflakeBase):
    """For parent review queue."""
    progress_id: int
    child_id: int
    child_name: str
    topic_id: int
    topic_name: str
    topic_description: str
    evidence: list[str]
    evidence_zh: list[str] | None
    attempts: int
    study_duration_seconds: int
    submitted_at: datetime
